packages/@lwc/engine-core/src/framework/base-lightning-element.ts

```diff
@@ -571,6 +571,24 @@ LightningElement.prototype = {
     },
 };
 
+function assertNotSyntheticComposedWithinNative(vm: VM) {
+    const isSynthetic =
+        vm.renderMode === RenderMode.Shadow && vm.shadowMode === ShadowMode.Synthetic;
+    if (!isSynthetic) {
+        return;
+    }
+    let ancestor = vm.owner;
+    while (!isNull(ancestor)) {
+        assert.isFalse(
+            ancestor.renderMode === RenderMode.Shadow && ancestor.shadowMode === ShadowMode.Native,
+            `<${vm.tagName}> (preferNativeShadow=false) cannot be composed inside of <${
+                getAssociatedVM(ancestor.elm).tagName
+            }> (preferNativeShadow=true)`
+        );
+        ancestor = ancestor.owner;
```

Comment:
This looks like we're checking all the component ancestors for each synthetic mode component, is that correct?

Reply:
Correct. Although we stop as soon as we hit a native shadow ancestor, since the `assert` will throw.

Reply:
I'll add a comment to make it clearer

Reply:
What I mentioned earlier is that I think each component only needs to check its immediate `RenderMode.Shadow` ancestor. If every synthetic mode component doesn't have a native mode owner, then this implies via transitivity that there are no native mode components that contain synthetic mode components.

Reply:
@ekashida Checking the immediate ancestor would work fine, except for light DOM. We might have a structure like native -> light -> light -> light -> synthetic. So in my algorithm, we keep climbing until we find a shadow component that is also native.

Reply:
Right, so for each synthetic mode component, we would climb up until we reach the first `RenderMode.Shadow` component, do a single assertion there, and stop. That seems like it might be a nice performance improvement.

Reply:
Maybe I'm misunderstanding what you're proposing. The code is:

https://github.com/salesforce/lwc/blob/16c3835c3c5683f55615d8bc0c1d0568ff006f22/packages/%40lwc/engine-core/src/framework/vm.ts#L261-L269

So you're suggesting this?

```js
     while (!isNull(ancestor)) { 
         if (ancestor.renderMode === RenderMode.Shadow) {
           assert.isFalse(
               ancestor.shadowMode === ShadowMode.Native, 
               `<${vm.tagName}> (preferNativeShadow=false) cannot be composed inside of <${ 
                   getAssociatedVM(ancestor.elm).tagName 
               }> (preferNativeShadow=true)` 
           ); 
         }
         ancestor = ancestor.owner; 
     } 
```

It doesn't seem like this would be a big perf improvement, unless it's very expensive to call `assert.isFalse()`?

Reply:
This is probably missing a null check somewhere but the idea is something along the lines of:
```
function isSyntheticShadow(vm) {
    return vm.renderMode === RenderMode.Shadow && vm.shadowMode === ShadowMode.Synthetic;
}
function isNativeShadow(vm) {
    return vm.renderMode === RenderMode.Shadow && vm.shadowMode === ShadowMode.Native;
}
function assertNotSyntheticComposedWithinNative(vm: VM) {
    if (!isSyntheticShadow(vm)) {
        return;
    }
    let ancestorVM = getClosestShadowModeAncestor(vm);
    assert.invariant(
        !isNativeShadow(ancestorVM),
        'blah blah blah'
    );
}
```

Reply:
I see, we are avoid the `assert` on light ancestors and native ancestors. OK, I'll do this.